### DISCUSSION TITLE
Can I directly modify the image data?

### DISCUSSION THREAD

**Discussion by xuyun018** [fltk/fltk#1030] [Development]
I will use the following code to draw the image.
//-------------------------------------------------
```c++
Fl_RGB_Image *pimage;

memset(pdata_, 255, 10000);
pimage = new Fl_RGB_Image(pdata_, width, height, 3);
pimage->draw(0, 0);
usleep(10);
memset(pdata_, 0, 10000);
pimage->draw(0, 0);
delete pimage;
```
//-------------------------------------------------
When I change the content of pdata_ and call the drawing function again, but the image remains unchanged, I think it is because FLTK has reconstructed the data of the image internally. Is there a way for me to directly obtain the image data and modify it to improve the speed?

**Comment by CyprinusCarpio** [comment]
The image data is cached by FLTK. I don't think you can modify it directly. Instead you need to un-cache the data using `Fl_RGB_Image::uncache` which will force FLTK to display the new data.
You can look at the example program I attached in issue #1023 which is a very simple raster painting program, that's using a `Fl_RGB_Image` to display the raster.

**Reply by xuyun018** [reply]
Thank you very much, it has been resolved.
If you can directly modify the buffer, please let me know.
Reducing memory usage and data replication will improve code efficiency, as I have a relatively large animation that needs to be implemented.

**Reply by MatthiasWM** [reply]
The cached image data is not stored as RGB or RGBA. It's stored in some native format, so that the undelying graphics library can copy pixels as efficiently as possible to the screen. This makes a lot of sense if the same image is drawn many times, typically during screen redraws. If you only draw RGB data once, see fl_draw_image(), as Erco already wrote.

**Reply by Albrecht-S** [reply]
See also this article: "[Mini-howto: Display Arbitrary Data as an Image](https://www.fltk.org/articles.php?L468)".

**Comment by erco77** [comment]
Have a look at:
https://www.seriss.com/people/erco/fltk/#Fl_RGB_Image_Inline_Example

This shows how to initialize an `Fl_RGB_Image` with an RGB buffer, and you should be able to access and change it using methods from the `Fl_Image` base class, e.g. data(), data_h(), data_w(), d(), count(), etc.

**Reply by erco77** [reply]
I should add, if you just want to dump a memory buffer's contents as pixels on the screen, you don't have to use instances of the `Fl_*_Image` classes, you can just call `fl_draw_image()` directly on the buffer within a `draw()` method derived from one of the common widget classes. In this example a custom class derived from `Fl_Window` to handle the `draw()` method calling fl_draw_image(): https://www.seriss.com/people/erco/fltk/#FlDrawImage
..but you can derive from other widgets too to implement your own `draw()`, like `Fl_Box` or `Fl_Widget`.
I think if you look at some of the other examples on that "FLTK cheat sheet" page, you'll see there's lots of ways to control drawing of images.

**Comment by xuyun018** [comment]
The ‘fl_draw_image’ function includes request (not always request) operations and data conversion operations.
I have written three functions on both Windows and Linux to meet this requirement.
